devito/propagator.py

```diff
@@ -235,36 +234,183 @@ def prep_variable_map(self):
+                cses[0].remove(cses[0][i])
+
+
+        cse_map=cse_map_update()
+        # unsub all indices
+        for i,item in enumerate(cses[0]):
+            left, right = item
+            if isinstance(right, Indexed):
+                index_replace={}
+                for i_index in list(right.indices):
+                    for i_replacements,replacement in enumerate(list(zip(*cses[0])[0])): # list of rhs replacements
+                        if(str(i_index) is str(replacement)):
+                            embed()
+                            index_replace[i_index]=cses[0][i_replacements][1]
+                            if str(replacement) not in list(cse_temp_vars.keys()):
+                                cse_temp_vars[str(replacement)] = self.dtype
+                            cse_map = cse_map_update()
+                cses[0][i]=(left,right.xreplace(index_replace))
+                for key in list(index_replace.keys()):
+                    left = key
+                    right = index_replace[key]
+                    # remove from temporaries
+                    cse_bad[cses[0][i][0]]=cses[0][i][1]
+                    cse_bad[left]=right
+                    cses[0].remove((left,right))
+                    cse_map = cse_map_update()
+
+
+        for i,item in enumerate(cses[0]):
+            left, right = item
+            if isinstance(right, Indexed):
+                assert(right.shape is not None)
+            elif isinstance(right, IndexedBase):
+                assert(right.shape is not None)
+
+        # x3=x1[x2,x,y,z]
+        # cses have had pointer aliases undone: x3=u[x2,x,y,z]
+        # cses have had indices unsubbed: x3=u[t-1,x,y,z]
+        # cses have had shapes fixed: x3.shape = (151,50,50,50)
+
+
+        for key in list(cse_bad.keys()):
+            symb = symbols(str(key))
+            cse_bad[symb] = cse_bad[key]
+
+            indexedbase = IndexedBase(str(key))
+            cse_bad[indexedbase] = cse_bad[key]
+
+        # go through cse'ed kernel list and rebuild missing shapes
+        kernel_sub_map={}
+        for idx,term in enumerate(cses[1]): # go through a list of new kernels
+            for arg in postorder_traversal(cses[1][idx]):
+                if isinstance(arg, IndexedBase):
+                    try:
+                        ccode(arg)
+                    except:
+                        print 'Failed to generate C code for %s' % arg
+                        embed()
+                elif isinstance(arg, Indexed):
+                    for i_index in list(arg.indices):
+                        for i_replacements,replacement in enumerate(list(zip(*cses[0])[0])): # list of rhs replacements
+                            if(str(i_index) is str(replacement)):
+                                if str(replacement) not in list(cse_temp_vars.keys()):
+                                    cse_temp_vars[str(replacement)] = 'int '
+                    try:
+                        ccode(arg)
```

Comment:
Is it possible to separate the code generation functionality from this? I think it would be much cleaner to only do "sympy expression" -> "list of sympy epxressions" in this routine and leave the code generation to Propagator or equivalent.